docs/5.x/terraform.md

```diff
@@ -0,0 +1,534 @@
+8wuHDWe7iv1ACCU6JeGjGJcKkpUTYl9UzMF5HCxLtPWHy9WFWX3EZtmbft40iA0h
+hlcrJ+NitQKBgQDMokHz8LEyKhEZOGoGsMxEAIvvbne2TXfTjR9VhOgkAE6lehuX
+ZYw77ue7D3rNCs/xPN/+cMmvyyGH1K5T+1itpz0f79uqTZkfLXqKODfrOa56Rdib
+LALJ8kqVNCkNFZmRKHUQqif8fqbtbKLaX0J0DdmS3bEiBVBB/lHptmTFCQKBgHB3
+y9g/vwfs/EaVDLUHhY8QpxBkz7033Bh+l0I16l8nCA+Vx6Xd6KRuAwIz4lip3OEX
+C7e3WeOPWBLTpzYuZjyAMasMG1CriGY70DiHAF/WYt1xAPLk0fmyEssa7M7uA9JI
+vBfZQsC23lZe3Tbc1LSOASvrl0HAN3nf/ANrfcLVAoGBAMD+5zcdxzrlueKAfH4i
+G4ShGDjaDXg/1HjZfi/x7Ur+umNaz868Wd7vSceQHuZlaf4tVmfrugjOBEMatSod
+sax3pz5xwntVlXbluLAC3lB3wm56L38eV/OD2Y21GHjqD1A0O6aZW/YwN8mjvCi4
+xDGW5r0S7Z891vy2DX4OJ9n3
+-----END PRIVATE KEY-----
+EOF
+}
+```
+
+### Argument Reference
+The following arguments are supported:
+
+* `cert` - The certificate chain in pem format.
+* `private_key` - The private key which matches the certificate in pem format.
+
+## gravity_token
+A token is a static secret that can be used to login to a cluster as a user.
+
+### Example Usage
+```
+resource "random_id" "secret_admin_token" {
+  byte_length = 32
+}
+
+resource "gravity_token" "test" {
+  token = "${random_id.secret_admin_token.hex}"
+  user  = "adminagent@example.com"
+}
+```
+
+### Argument Reference
+The following arguments are supported:
+
+* `token` - A secret token that can be used to access the cluster.
+* `user` - The user the token is for.
+
+## gravity_user
+A local cluster user
+
+### Example Usage
+```
+resource "random_id" "admin_password" {
+  byte_length = 32
+}
+
+resource "gravity_user" "test" {
+  name      = "test"
+  full_name = "Test User"
+  type      = "admin"
+  roles     = ["@teleadmin"]
+  password  = "${random_id.admin_password.hex}"
+}
+```
+
+### Argument Reference
+The following arguments are supported:
+
+* `name` - An email address style username.
+* `full_name` - (Optional) A friendly name for the user.
+* `password` - (Optional) A password to provision for the user.
+* `roles` - A customized list of roles.
+
+
+# Terraform Provider (Enterprise)
+The gravity enterprise terraform provider is used to support terraform management of resources only available in the enterprise version of gravity. This provider should be used in conjunction with the opensource gravity provider to manage a gravity cluster.
+
+## Getting Started
+
+### Install the Gravity Enterprise provider
+The terraform provider will be automatically installed when getting the gravity tools.
+
+```
+curl https://get.gravitational.io/telekube/install/5.0.23 | bash
+```
+
+Please see the [getting started guide](https://gravitational.com/telekube/docs/quickstart/#getting-the-tools) for more information.
+
+### Example Usage
+
+```
+# Configure the Gravity provider
+provider "gravity" {
+    host  = "https://example.com"
+    token = "abcdefghi"
+}
+
+# Create a log forwarder
+resource "gravity_log_forwarder" "logs" {
+    # ...
+}
+
+# Configure the gravity enterprise provider
+provider "gravityenterprise" {
```

Comment:
Hmm, I don't understand why we have a separate provider and prefix for resource names for enterprise. Why can't it just be "gravity" everywhere and whether a certain resource is supported or not will depend on which edition of terraform-provider-gravity it is run with (oss or ent). I.e. if you build enterprise binary "terraform-provider-gravity", it will recognize "gravity_oidc", but the oss one will not.

Reply:
Well, I think this model of one binary that get's built differently will be problematic in the long run. I see a few weaknesses with this approach:
- If the resources are simply missing in the OSS edition, it will throw very generic configuration errors, which could be especially confusing since docs would show this to be valid configuration. It would be possible to stub this and provide better errors, but to me it's still something that will lead to confusion. Configuration or examples from the docs, will simply not work because the wrong version of the module is installed.
- If a user starts with OSS, begins adopting enterprise, and wants to manage clusters of both types, they could inadvertently use the wrong module with the wrong cluster, again leading to confusion.
- If the edition isn't in the filename, it's difficult to look at the file on the system and see which version is present. Terraform uses the provider name to locate the file, so how is a user to figure out how to correct the situation.

So because we have a very clear separation in which API endpoints are for oss/enterprise, the modules were built to mirror this, and named accordingly.